File: leetCode/5.普通数组/56合并区间.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 * 以数组 intervals 表示若干个区间的集合，其中单个区间为 intervals[i] = [starti, endi] 。
 * 请你合并所有重叠的区间，并返回 一个不重叠的区间数组，该数组需恰好覆盖输入中的所有区间 。
 * 举例：intervals = [[1,3],[2,6],[8,10],[15,18]]
 * 答案：[[1,6],[8,10],[15,18]]
 * 区间 [1,3] 和 [2,6] 重叠, 将它们合并为 [1,6].
 */
var merge = function(intervals) {
    if (intervals.length == 0) return []
    const sortIntervals = intervals.sort((a, b) => a[0] - b[0])
    const answer = [intervals[0]]
    for(let i = 1; i < sortIntervals.length; i++) {
        const lastItem = answer.pop()
        const item = sortIntervals[i]
        if (lastItem[1] >= item[0] && lastItem[1] < item[1]) {
            lastItem[1] = item[1]
        }
        if (lastItem[1] < item[0]) {
            answer.push(lastItem)
            answer.push(item)
        } else {
            answer.push(lastItem)
        }
         
    }
    return answer
}
console.log(merge([[1,3],[2,6],[8,10],[15,18]]))

File: leetCode/5.普通数组/56合并区间.go
func merge(intervals [][]int) [][]int {
    sortIntervals :=  sort(intervals)
    window := sortIntervals[0]
    result := [][]int{}
    for i := 1; i < len(sortIntervals); i++ {
        if (window[1] >= sortIntervals[i][0]) {
            if window[1] >= sortIntervals[i][1] {
                continue
            } else {
                window[1] = sortIntervals[i][1]
            }
        } else {
            result = append(result, window)
            window = sortIntervals[i]
        }
    }
    result = append(result, window)
    return result
}
func sort(nums [][]int) [][]int {
    if len(nums) == 0 {
        return nil
    }
    val := nums[0]
    leftArray := [][]int{}
    rightArray := [][]int{}

    for i := 1; i < len(nums); i++ {
        if nums[i][0] <= val[0] {
            leftArray = append(leftArray, nums[i])
        } else {
            rightArray = append(rightArray, nums[i])
        }
    }
    return append(append(sort(leftArray), val), sort(rightArray)...)

}